pkg/distributor/distributor.go

```diff
@@ -640,6 +655,28 @@ func (d *Distributor) wrapPushWithMiddlewares(next push.Func) push.Func {
 	return next
 }
 
+// prePushUserRequestMiddleware increments a per-tenant request counter before other middlewares run
+// (HA deduplication, forwarding) so that we can get an accurate idea of number of requests per user.
+func (d *Distributor) prePushUserRequestMiddleware(next push.Func) push.Func {
+	return func(ctx context.Context, req *mimirpb.WriteRequest, cleanup func()) (*mimirpb.WriteResponse, error) {
+		cleanupInDefer := true
+		defer func() {
+			if cleanupInDefer {
+				cleanup()
+			}
+		}()
+
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		cleanupInDefer = false
+		d.incomingRequests.WithLabelValues(userID).Add(1)
```

Comment:
This line is literally the only required change in business logic. Shall we include it in `d.instanceLimitsMiddleware` instead? (We can rename that middleware, if that bothers you.)

Reply:
Yeah, it's the only line. Sure, I can add it to the instance limits middleware.

Reply:
Updated in 1777b9caca9f751b065d162b7b678bf2bf92d88b